src/librustc_trans_utils/trans_crate.rs

```diff
@@ -77,179 +77,28 @@ pub trait TransCrate {
-            trans_crate.provide_extern(providers);
-        }
-
-        #[no_mangle]
-        pub extern "C" fn __rustc_backend_trans_crate<'a, 'tcx: 'a>(
-            trans_crate: *const Box<TransCrate>,
-            tcx: *mut TyCtxt<'a, 'tcx, 'tcx>,
-            rx: *mut mpsc::Receiver<Box<Any + Send>>
-        ) -> *mut Box<Any> {
-            let trans_crate = unsafe { &*trans_crate };
-            let tcx = unsafe { *tcx };
-            let rx = unsafe { *Box::from_raw(rx) };
-            let trans = trans_crate.trans_crate(tcx, rx);
-            Box::into_raw(Box::new(trans))
-        }
-
-        #[no_mangle]
-        pub extern "C" fn __rustc_backend_join_trans_and_link(
-            trans_crate: *const Box<TransCrate>,
-            trans: *mut Box<Any>,
-            sess: *const Session,
-            dep_graph: *const DepGraph,
-            outputs: *const OutputFilenames
-        ) -> *mut Result<(), CompileIncomplete> {
-            let trans_crate = unsafe { &*trans_crate };
-            let trans = unsafe { *Box::from_raw(trans) };
-            let sess = unsafe { &*sess };
-            let dep_graph = unsafe { &*dep_graph };
-            let outputs = unsafe { &*outputs };
-            let result = trans_crate.join_trans_and_link(trans, sess, dep_graph, outputs);
-            Box::into_raw(Box::new(result))
-        }
-    }
-}
-
-pub struct ExternTransCrate {
-    lib: ::libloading::Library,
-    backend: Box<Box<TransCrate>>,
-}
-
-macro_rules! get_symbol {
-    (($lib:expr) . $name:ident : $type:ty) => {
-        let $name: ::libloading::Symbol<$type> = $lib.get(stringify!($name).as_bytes()).unwrap();
     }
 }
 
-impl ExternTransCrate {
-    pub fn new<P: AsRef<OsStr>>(sess: &Session, filename: P) -> Box<TransCrate> {
-        use libloading::*;
-        let filename = filename.as_ref();
-        match Library::new(filename) {
-            Ok(lib) => {
-                let backend = unsafe {
-                    get_symbol!((lib).__rustc_backend_new:
-                        unsafe extern "C" fn(&Session) -> *mut Box<TransCrate>);
-                    Box::from_raw(__rustc_backend_new(sess))
-                };
-                Box::new(ExternTransCrate {
-                    lib,
-                    backend,
-                })
-            }
-            Err(err) => {
-                sess.fatal(&format!("Couldnt load codegen backend {:?}: {:?}", filename, err));
+pub fn link_extern_backend<P: AsRef<OsStr>>(sess: &Session, filename: P) -> Box<TransCrate> {
+    use libloading::*;
+    let filename = filename.as_ref();
+    match Library::new(filename) {
+        Ok(lib) => {
+            unsafe {
+                let __rustc_backend_new: Symbol<unsafe fn(&Session) -> Box<TransCrate>>;
+                __rustc_backend_new = lib.get(b"__rustc_backend_new")
+                    .expect("Couldnt load codegen backend as it\
+                    doesnt export the __rustc_backend_new symbol");
```

Comment:
Can't you match on this to use `sess.fatal` here? Also, missing `'` in `Couldn't`.